third_party/maya/lib/usdMaya/MayaParticleWriter.cpp

```diff
@@ -0,0 +1,362 @@
+    const TfToken _massName("mass");
+
+    template <typename T>
+    void _addAttrVec(UsdGeomPoints& points, const SdfValueTypeName& typeName, const _strVecPairVec<T>& a,
+                     const UsdTimeCode& usdTime) {
+        for (const auto& v : a) {
+            _addAttr(points, v.first, typeName, *v.second, usdTime);
+        }
+    }
+
+    // The logic of filtering the user attributes is based on partio4Maya/PartioExport.
+    // https://github.com/redpawfx/partio/blob/redpawfx-rez/contrib/partio4Maya/scripts/partioExportGui.mel
+    // We either don't want these or already export them using one of the builtin functions.
+    const std::set<std::string> _supressedAttrs = {
+        "emitterDataPosition", "emitterDataVelocity", "fieldDataMass",
+        "fieldDataPosition", "fieldDataVelocity", "inputGeometryPoints",
+        "lastCachedPosition", "lastPosition", "lastVelocity",
+        "lastWorldPosition", "lastWorldVelocity", "worldVelocityInObjectSpace",
+        "position", "velocity", "acceleration", "rgb", "rgbPP", "incandescencePP",
+        "radius", "radiusPP", "age", "opacity", "opacityPP", "lifespan", "lifespanPP",
+        "id", "particleId", "mass",
+    };
+
+    // All the initial state attributes end with 0
+    bool _isInitialAttribute(const std::string& attrName) {
+        return attrName.back() == '0';
+    }
+
+    bool _isCachedAttribute(const std::string& attrName) {
+        constexpr auto _cached = "cached";
+        constexpr auto _Cache = "Cache";
+        const auto attrNameSize = attrName.size();
+        if (attrNameSize > 5 && attrName.substr(attrNameSize - 6) == _Cache) {
+            return true;
+        } else {
+            return attrName.substr(0, 6) == _cached;
+        }
+    }
+
+    bool _isValidAttr(const std::string& attrName) {
+        if (attrName.size() == 0) { return false; }
+        if (_isInitialAttribute(attrName)) { return false; }
+        if (_isCachedAttribute(attrName)) { return false; }
+        return _supressedAttrs.find(attrName) == _supressedAttrs.end();
+    }
+}
+
+MayaParticleWriter::MayaParticleWriter(
+    const MDagPath & iDag,
+    const SdfPath& uPath,
+    bool instanceSource,
+    usdWriteJobCtx& jobCtx)
+    : MayaTransformWriter(iDag, uPath, instanceSource, jobCtx),
+      mInitialFrameDone(false) {
+    auto primSchema = UsdGeomPoints::Define(getUsdStage(), getUsdPath());
+    TF_AXIOM(primSchema);
+    mUsdPrim = primSchema.GetPrim();
+    TF_AXIOM(mUsdPrim);
+
+    initializeUserAttributes();
+}
+
+void MayaParticleWriter::write(const UsdTimeCode &usdTime) {
+    UsdGeomPoints primSchema(mUsdPrim);
+    writeTransformAttrs(usdTime, primSchema);
+    writeParams(usdTime, primSchema);
+}
+
+void MayaParticleWriter::writeParams(const UsdTimeCode& usdTime, UsdGeomPoints& points) {
+    if (usdTime.IsDefault() == isShapeAnimated()) {
+        return;
+    }
+
+    const auto particleNode = getDagPath().node();
+    MFnParticleSystem PS(particleNode);
```

Comment:
Just one nitpick: for consistency with coding conventions, could we change the variable names ``PS`` and ``DPS``?  I think i would've gone with ``particleSys`` and ``deformedParticleSys`` but that is quite a lot to type..